### PR TITLE
THRIFT-4373: Derefer PHP zval _TSPEC

### PR DESCRIPTION
**Pull request by sokac**

### CODE REVIEW

File: lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp [hunk 1]
```diff
@@ -537,6 +537,9 @@ void binary_deserialize(int8_t thrift_typeID, PHPInputTransport& transport, zval
       }
 
       zval* spec = zend_read_static_property(Z_OBJCE_P(return_value), "_TSPEC", sizeof("_TSPEC")-1, false);
+      if (spec) {
```
Comment (RobberPhex) [inline lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp]: After `zend_read_static_property` with `slient=false`, we should detect exception.
If no exception, we needn't detect spec is null.

but we need detect that `Z_TYPE_P(prop) == IS_REFERENCE`

File: lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp [hunk 2]
```diff
@@ -699,6 +702,9 @@ void binary_serialize_hashtable_key(int8_t keytype, PHPOutputTransport& transpor
 
 static
 void binary_serialize(int8_t thrift_typeID, PHPOutputTransport& transport, zval* value, HashTable* fieldspec) {
+  if (value) {
```
Comment (RobberPhex) [inline lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp]: `Z_TYPE_P(prop) == IS_REFERENCE`

File: lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp [hunk 3]
```diff
@@ -709,6 +715,9 @@ void binary_serialize(int8_t thrift_typeID, PHPOutputTransport& transport, zval*
         throw_tprotocolexception("Attempt to send non-object type as a T_STRUCT", INVALID_DATA);
       }
       zval* spec = zend_read_static_property(Z_OBJCE_P(value), "_TSPEC", sizeof("_TSPEC")-1, true);
+      if (spec) {
```
Comment (RobberPhex) [inline lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp]: I think `Z_TYPE_P(prop) == IS_REFERENCE`

**Comment by RobberPhex** [comment]
If we want process REFERENCE everywhere.
I think we should review every line of `zval` used, is it?

File: lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp (hunk 1 above)
Comment (sokac) [reply lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp]: ZVAL_DEREF checks is reference:
https://github.com/php/php-src/blob/ffc734b2a455e2f2748f18a54ab597f7e0c715ba/Zend/zend_types.h#L944

I removed if statement

File: lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp (hunk 2 above)
Comment (sokac) [reply lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp]: removed if statement

File: lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp (hunk 3 above)
Comment (sokac) [reply lib/php/src/ext/thrift_protocol/php_thrift_protocol.cpp]: removed if statement

**Comment by sokac** [comment]
@RobberPhex how does this look now? I don't have lots of context for other zval values so I'd leave that for another review.

**Comment by james-johnston-thumbtack** [comment]
@RobberPhex  Any word on getting this merged?  Using the extension in PHP 7.0 is kinda broken otherwise... this is fixing a common issue that affected many PHP extensions moving to PHP 7.0.

**Comment by jeking3** [comment]
Need to see if pass the CI build - please squash to a single commit, rebase on master, and force push.

**Comment by sokac** [comment]
@jeking3 squashed and rebased. Looks like CI has some troubles, though

**Pull request merged**